### ISSUE TITLE
Fix recording of audio via Ctrl-Alt-F5 on PPC

### ISSUE THREAD

**Issue by kas1e** [dreamer/dosbox-staging#161]
Tested with latest DosBox-Staging and with latest original DosBox repo: in all cases, recorded video have a distorted sound if you record anything which has sound/music.

**Comment by dreamer** [comment]
Sooo... upstream just made a massive change in video recording and screenshot captures. And I mean massive, they almost rewrote the whole thing. Instead of 4 line fix as I did, the whole change is: 6 files changed, 662 insertions(+), 370 deletions(-); I think this rewrite was brewing for some time in private, they just released it now.

On a negative side, it means I need to resolve conflicts **and we'll need to retest the big-endian related issues**. On a positive side - their version might fix audio capture. jmarsh forgot to document new config options in INSTALL file, so I presume more changes might arrive shortly in SVN.

I'll prepare test branch merging trunk into our master branch later today, stay tuned.

**Comment by dreamer** [comment]
@kas1e can you test branch `po/merge-test-r4314` in regards to audio recording on PPC AND screenshots use-case?

**Comment by kas1e** [comment]
@dreamer 
Tested: yes, video recording works now and produce correct video and output. And screenshots at least in 16 bit save correctly.

But the recording of audio somehow "interrupts". I.e. I even tried to play in some simple game which didn't load CPU, and still video records with interrupted audio every few seconds. Is it the same for you on x86? Maybe it because of cycles "max" and that this version are without PPC-jit...  Will check more.

**Comment by kas1e** [comment]
Well I even set CPU cycles to 5000, and when playing in-game it just shows my CPU loading is about 3-5% only, and still, video saves with interrupted audio. 

Will check also last code from main repo, but something makes me think it will be the same...

**Comment by kas1e** [comment]
Build win32 (MinGW) versions of original DosBox and of a branch of staging we test now: in both cases, the audio didn't interrupt in a recorded video on my x86. In both cases, it, of course, built with x86 jit, so maybe that somehow related...

**Comment by krcroft** [comment]
@kas1e , if you can zip or tar the recordings, you can attach them here - limit is 20MB/file, which beats vogons.  Would be great for jmarsh to see the comparisons (assuming he did the re-write). 

Very interesting the encoder isn't keeping up with realtime (resulting in stuttering being emebedded in the recordings), despite CPU usage is low.

**Comment by dreamer** [comment]
There are several weird things happening here (and I haven't looked at the audio issue yet):

- Change from jmarsh is missing documentation in INSTALL file (ok… upstream does not care about having up-to-date documentation anyway, not a big deal… - but it's another undocumented compilation switch - same situation as with `--disable-screenshots`, except I don't see the reason for a new build configuration in this case. @kas1e, in theory, we should add it as another build config to test with our "Config heavy" workflow, but I'm not sure if we want it.
- Change does not compile using new MSVC (working on a fix)
- For some strange reason, feature of taking screenshots seems to be magically turned off on Clang on macOS (no idea why, everything seems ok in autoconf log, all the dependencies are in…)

It's annoying, that such large change was dumped in one big commit instead as a series of small patches :(

**Comment by krcroft** [comment]
~1,000 lines of code are committed _in one fell-swoop_ to core DOSBox yet without a single issue being flagged, patch being posted, or feature request being made on DOSBox's own official project website. 

![2020-02-06_20-04_1](https://user-images.githubusercontent.com/1557255/74000301-6758b480-491c-11ea-9aa8-14b35006703f.png)
![2020-02-06_20-04](https://user-images.githubusercontent.com/1557255/74000300-66c01e00-491c-11ea-9eda-8af421c56e43.png)
![2020-02-06_20-04_2](https://user-images.githubusercontent.com/1557255/74000302-6758b480-491c-11ea-9807-bed5b61650f1.png)

I'm not sure why such sweeping changes need to be planned, tested, and agreed upon in private without any discussion or involvement from the users or community.

**Comment by dreamer** [comment]
Just tested new video capture on Windows 10 (while fixing the compilation issue); audio seems fine (tested using Retro City Rampage and Spear of Destiny), but there's no video at all (VLC complains about missing codec).

Before this change, I was able to play videos recorded using DOSBox or dosbox-staging just fine, I didn't need to install zmbv "codec" system-wide. And I'm not sure if I even want to install it...

Not sure how to proceed yet... I really don't want to merge this SVN revision with all the problems... I'll do some more testing.

**Comment by kas1e** [comment]
@dreamer 
Yes, for me on win32 the same issue: audio writes fine (no interrupts), but no video fully.

**Comment by kas1e** [comment]
@all
https://www.vogons.org/viewtopic.php?f=32&t=71837

**Comment by dreamer** [comment]
And tested recording on Linux (never tried it before); I needed to manually change shortcut in mapper (Ctrl-Alt-F5 changes to tty5… I think I'll change this shortcut to Ctrl-Shift-F5 by default in dosbox-staging in some other commit).

On Linux I get the same broken behaviour using DOSBox 0.74-3 and our testing branch: video records fine, but there is no sound at all.

**Comment by kas1e** [comment]
@dreamer 
Probably we need to discuss it on vogons (so jmarsh will see it), as he probably not follow anything else :)

**Comment by dreamer** [comment]
I don't think a discussion on Vogons will be productive. We already heard, the usual stuff: response to you is "your OS is broken", response to me is "your VLC is broken", my comment was dismissed. Such discussions tend to make my blood boil, so I will probably refrain from further comments in there.

Whenever we had such discussions on Vogons, after we heard how stupid and incompetent we are, the fix landed in SVN a day or two later - I guess that'll be the pattern this time as well.

I will refrain from merging r4314 into master branch for now and focus on other tasks; if upstream won't fix the issue in few days, I'll look into developing our own fix to audio recording issue, so we could revert r4314 while merging.

**Comment by krcroft** [comment]
ffmpeg in analysis-mode (`ffmpeg -v verbose -i file.avi -f null -`) reports having to guess audio stream layout; however, this isn't an error -- but perhaps that subtly is giving other players issues?

When I feed other AVIs into ffmpeg, it doesn't have to guess audio stream layout; so that's the one difference I noticed. Click to see in fullscreen (sorry for the screenshot.. easier to read than a wall of gray text)... 

![2020-02-06_23-26](https://user-images.githubusercontent.com/1557255/74009835-ec51c700-4938-11ea-85f5-ac2bcf79236e.png)

**Comment by kas1e** [comment]
I made post on vogons only so he will aware, but "your vlc is broken" answet are such double-facepalm!

**Comment by krcroft** [comment]
Added the one-liner, ` --disable-recording` to config-heavy under `kc/r4314-config-heavy`; when this video rework effort is ready.

**Comment by dreamer** [comment]
New commits landed in SVN, but nothing related to video recording.

@kas1e Is ffmpeg available on AmigaOS? I think we should consider dropping DOSBox-specific ZMBV codec and replacing it with something simpler (maybe using one of AV-1 encoders running in a separate thread or developing a way to dump a stream of raw frames and encode them on the fly using ffmpeg).

**Comment by kas1e** [comment]
@dreamer
ffmpeg available, but it may take more cpu time in recording imho (?) which is important when you record heavy games.. 

I will check tomorrow if installing ZMBV codec will help on my win10. Also will dbl-recheck if i record video with stopping (as some point out if i just close dosbox it then may be broken).

Another strange issue is that on vogons some user says that he can play my file without audio issues, which mean that maybe video recording produce broken video by some reassons

**Comment by dreamer** [comment]
I think that comment about stopping recording before closing DOSBox is just stupid - DOSBox should take care to correctly stop recording and close file on exit. Or *at least* use a temporary name until the recording is stopped to accommodate possible crashes and such.

I don't know if using ffmpeg will be slower - yes, it is more computationally intensive, but devs spent many man-years optimizing it and developing modern codecs with good encoding speed, as opposed to zmbv, which is used only by DOSBox and no other projects. Putting encoding on a separate thread should alleviate problems with recording CPU-intensive games. So: we don't know, we would need to test it ;)

**Comment by krcroft** [comment]
It makes a lot of sense to separate the tasks. First, the encoding is currently done inside the emulation loop; between each video frame you've got a portion of DOSBox code being spent into encoding (which directly competes with the core emulation).  

Moving the encode to a separate process means DOSBox gets to run at its original speed, plus you now have OS-scheduler control over the encode process (so you can can run it at a lower priority).  Even on a single-core CPU, you can run the encoder at a very low-priority (idle-cycles) and with a big-enough receive buffer, it won't drop frames either - could even spool to disk; means the encoder could still be running for quite some time /after/ DOSBox closes, with enough cycle available to digest the queue.

Another benefit is disconnecting technologies.  Video and audio compression continue to evolve, and this lets users capture using whatever codecs suit their needs.  It's better for maintainability - DOSBox developer want to emulate DOS; they don't want to be video and audio codec maintainers; leave that to the ffmpeg folks!

I suppose the challenge is now figuring out how to crank out frames and PCM audio via whatever cross-platform interface is best suited to streaming capture.  We could even use stdout; `dosbox | ffmpeg ... `!

**Comment by krcroft** [comment]
>  some user says that he can play my file without audio issues

Yes, it plays smooth for me too.  To get a %100 independent assessment, use a completely different system and playback software.  For example, on my Android phone, I played the video using MX Player, a lean and fast open-source player (that yes, even supports the dosbox codec :-) )

https://play.google.com/store/apps/details?id=com.mxtech.videoplayer.ad&hl=en_US

**Comment by dreamer** [comment]
@kas1e Can you test `po/fix-audio-record-ppc-1` branch on PPC? I believe it should have fixed audio, but otherwise behave like DOSBox before r4314.

When testing recording I found other issues affecting video recording since 0.74-3 (and not fixed by r4314): e.g. recording of HUGI results in broken video when playing in Totem. The problem with VLC not closing also happens since 0.74-3.

**Comment by kas1e** [comment]
@dreamer 

> Can you test ```po/fix-audio-record-ppc-1``` branch on PPC? I believe it should have fixed audio, but otherwise behave like DOSBox before r4314.

Yeah! The sound recording is fine now with your change!

> I think that comment about stopping recording before closing DOSBox is just stupid - DOSBox should take care to correctly stop recording and close file on exit. Or at least use a temporary name until the recording is stopped to accommodate possible crashes and such.

What that man who wrote this mean, is that DosBox has a bug: if you didn't stop video recording by pressing second time "ctrl+alt+f5", but just exit from DosBox, then a video will be broken, and that bug there for years. 

But there is anyway something wrong happens.  The recorded video on PPC when I tried to play it in default win10 player still show nothing, but in VLC it seems to show the things, but (!) in wrong colors. So or that DOSBox's codec is non-endian conform (i.e. videos written on big-endian will be wrong looking on little-endian) or there is a bug. 

Also, no seeking working when I load it up in a win10 default player (and have only music) meaning that file indeed is broken.

**Comment by dreamer** [comment]
I went back to testing on Windows 10 and can't reproduce my previous problems any more... I still can't play in Windows default player, but recordings work now in VLC... :/

But I found new bugs: colour recording in Retro City Rampage is wrong - all the magenta is replaced with green. Also, some frames seem to be missing (there are some strobing animations in RCR that are not recorded at all) - but these seem to be not regressions, as I can see the same issues in the snapshot build from master branch.

I am lost and not sure how to proceed - there are obvious bugs with video recording in DOSBox, but I'm not sure they are regressions any more - zmbv coded seems to be generally crappy and works kind-of-ok only in VLC for me now. There is an issue of interrupted audio you hear (only on PPC, right?) - I can't reproduce it, but after looking at the code I wouldn't be surprised if there was an audio artifact, only on PPC, introduced every 1024 samples.

**Comment by kas1e** [comment]
Ok, I made some big-fat tests, so to avoid recopy it there check plz Vogon's thread. But in brief, i tested r4313 and r4314 on both PPC and x86, as a result of those tests, I can say that:

1). There are bugs in a video recording on any platform. BSPlayer says something broken, VLC produces wrong colors, default win10 player fails to show video (which ZMBV codec installed).
That happens and when I record file on PPC, and when I record file on X86. Does not matter if it r4313 or 4314.

2). In terms of Big-Endian (at least on PPC and AmigaOS4), with r4314 things start to be worse: it seems, that building of Index fail now or something, while with 4313 it was ok (still with all issues I mentioned in 1). ).

3). "audio-interrupt" seems was an issue on one of PPC players, while in Mplayer port all fine. And those "audio-interrupts" happens does not matter with what revisions and platform video recorded, so I think its player's bug (or ZMBV is that crapy-shit that almost no one can play it all correctly).

**Comment by dreamer** [comment]
Issue got closed because I merged the fix with byteswap on PPC we tested earlier; I reverted r4314 and r4315 and merged later SVN revisions into master.

As for other bugs related to video recording - we can put them into their dedicated issues, but frankly speaking I don't feel like fixing dosbox-specific codec. I would rather remove zmbv encoding from dosbox-staging, and implement dumping raw frames into Y4M container, that could be then redirected to separate ffmpeg process (probably via named pipe), or e.g. a AV1 encoder.
